// @gnommostudios/ng-gnommo-base/esm5/lib/gnommo-base/services/core/base.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { map } from 'rxjs/operators';
import { Injectable, Inject } from '@angular/core';
import { HttpHeaders, HttpClient } from '@angular/common/http';
// unsupported: template constraints.
/**
 * @abstract
 * @template T
 */
var BaseService = /** @class */ (function () {
    function BaseService(http, url, endpoint) {
        this.http = http;
        this.url = url;
        this.endpoint = endpoint;
        this.httpOptions = {
            headers: new HttpHeaders({
                // tslint:disable-next-line:max-line-length
                'Accept': '*',
                'Content-type': 'application/json',
            })
        };
    }
    /**
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.getAll = /**
     * @param {?=} headers
     * @return {?}
     */
    function (headers) {
        return this.http.get(this.url + "/" + this.endpoint, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.getById = /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    function (_id, headers) {
        return this.http.get(this.url + "/" + this.endpoint + "/" + _id, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.create = /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    function (item, headers) {
        return this.http.post(this.url + "/" + this.endpoint, item, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.update = /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    function (_id, item, headers) {
        item._id = _id;
        return this.http.patch(this.url + "/" + this.endpoint, item, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.updateComplete = /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    function (_id, item, headers) {
        item._id = _id;
        return this.http.put(this.url + "/" + this.endpoint, item, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.delete = /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    function (_id, headers) {
        return this.http.delete(this.url + "/" + this.endpoint + "/" + _id, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.createHttpHeaders = /**
     * @param {?=} headers
     * @return {?}
     */
    function (headers) {
        /** @type {?} */
        var httpOptions = Object.assign({}, this.httpOptions);
        if (headers) {
            Object.keys(headers).forEach(function (headerKey) {
                httpOptions.headers = httpOptions.headers.append(headerKey, headers[headerKey]);
            });
        }
        return httpOptions;
    };
    /**
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.httpHeadersWithoutAuth = /**
     * @param {?=} headers
     * @return {?}
     */
    function (headers) {
        /** @type {?} */
        var httpOptions = {
            headers: new HttpHeaders({
                'Accept': '*',
                'Content-type': 'application/json',
                'X-Skip-Interceptor': ''
            })
        };
        if (headers) {
            Object.keys(headers).forEach(function (headerKey) {
                httpOptions.headers = httpOptions.headers.append(headerKey, headers[headerKey]);
            });
        }
        return httpOptions;
    };
    BaseService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    BaseService.ctorParameters = function () { return [
        { type: HttpClient, decorators: [{ type: Inject, args: [HttpClient,] }] },
        { type: String, decorators: [{ type: Inject, args: ['url',] }] },
        { type: String, decorators: [{ type: Inject, args: ['endpoint',] }] }
    ]; };
    return BaseService;
}());
export { BaseService };
if (false) {
    /** @type {?} */
    BaseService.prototype.httpOptions;
    /** @type {?} */
    BaseService.prototype.http;
    /** @type {?} */
    BaseService.prototype.url;
    /** @type {?} */
    BaseService.prototype.endpoint;
}
