// @gnommostudios/ng-gnommo-base/esm5/lib/gnommo-base/services/core/login-base.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { BaseService } from './base.service';
// unsupported: template constraints.
/**
 * @template T
 */
var LoginBaseService = /** @class */ (function (_super) {
    tslib_1.__extends(LoginBaseService, _super);
    function LoginBaseService(http, url, endpoint) {
        var _this = _super.call(this, http, url, endpoint) || this;
        _this.http = http;
        _this.url = url;
        _this.endpoint = endpoint;
        return _this;
    }
    /**
     * @param {?} credentials
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.login = /**
     * @param {?} credentials
     * @param {?=} headers
     * @return {?}
     */
    function (credentials, headers) {
        return this.http.post(this.url + "/" + this.endpoint + "/login", credentials, this.httpHeadersWithoutAuth(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} email
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.passwordRecovery = /**
     * @param {?} email
     * @param {?=} headers
     * @return {?}
     */
    function (email, headers) {
        return this.http.post(this.url + "/" + this.endpoint + "/password-recovery", email, this.httpHeadersWithoutAuth(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} newPassword
     * @param {?} hash
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.resetPassword = /**
     * @param {?} newPassword
     * @param {?} hash
     * @param {?=} headers
     * @return {?}
     */
    function (newPassword, hash, headers) {
        return this.http.post(this.url + "/" + this.endpoint + "/reset-password", { 'newPassword': newPassword, 'hash': hash }, this.httpHeadersWithoutAuth(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} token
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.getCurrentUser = /**
     * @param {?} token
     * @param {?=} headers
     * @return {?}
     */
    function (token, headers) {
        return this.http.get(this.url + '/users-me', this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.register = /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    function (item, headers) {
        return this.http.post(this.url + "/" + this.endpoint + "/register", item, this.httpHeadersWithoutAuth(headers))
            .pipe(map(function (data) { return data; }));
    };
    LoginBaseService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    LoginBaseService.ctorParameters = function () { return [
        { type: HttpClient, decorators: [{ type: Inject, args: [HttpClient,] }] },
        { type: String, decorators: [{ type: Inject, args: ['url',] }] },
        { type: String, decorators: [{ type: Inject, args: ['endpoint',] }] }
    ]; };
    return LoginBaseService;
}(BaseService));
export { LoginBaseService };
if (false) {
    /** @type {?} */
    LoginBaseService.prototype.http;
    /** @type {?} */
    LoginBaseService.prototype.url;
    /** @type {?} */
    LoginBaseService.prototype.endpoint;
}
