// @gnommostudios/ng-gnommo-base/esm2015/lib/gnommo-base/services/core/auth.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { CookieStorage } from '../../storage/storage.handler';
import { AuthToken } from '../../models/auth-token.model';
/** @type {?} */
const TOKEN_ID = 'token_id';
/** @type {?} */
const TOKEN_TTL = 'token_ttl';
/** @type {?} */
const TOKEN_CREATED = 'token_created';
export class AuthService {
    /**
     * @param {?} storage
     */
    constructor(storage) {
        this.storage = storage;
        this.token = new AuthToken();
        this.token.id = this.storage.get(TOKEN_ID);
        this.token.ttl = this.storage.get(TOKEN_TTL);
        this.token.created = this.storage.get(TOKEN_CREATED);
    }
    /**
     * @return {?}
     */
    getToken() {
        return /** @type {?} */ (this.token);
    }
    /**
     * @param {?} tokenId
     * @return {?}
     */
    setToken(tokenId) {
        /** @type {?} */
        const token = new AuthToken();
        token.id = tokenId;
        this.persist(TOKEN_ID, token.id, this.expiresTime());
        this.persist(TOKEN_TTL, this.expiresTime(), this.expiresTime());
        this.persist(TOKEN_CREATED, new Date(), this.expiresTime());
        token.ttl = this.expiresTime();
        token.created = new Date();
        this.token = Object.assign({}, this.token, token);
    }
    /**
     * @return {?}
     */
    removeToken() {
        this.storage.remove(TOKEN_ID);
        this.storage.remove(TOKEN_TTL);
        this.storage.remove(TOKEN_CREATED);
        this.token = new AuthToken();
    }
    /**
     * @param {?} token_property
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    persist(token_property, value, expires) {
        try {
            this.storage.set(`${token_property}`, (typeof value === 'object') ? JSON.stringify(value) : value, expires ? expires : null);
        }
        catch (err) {
            // console.error('Cannot access local/session storage:', err);
        }
    }
    /**
     * @return {?}
     */
    expiresTime() {
        /** @type {?} */
        const today = new Date();
        return this.addDays(today, 3);
    }
    /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    addDays(date, days) {
        /** @type {?} */
        const result = new Date(date);
        result.setDate(result.getDate() + days);
        return result;
    }
}
AuthService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
AuthService.ctorParameters = () => [
    { type: CookieStorage, decorators: [{ type: Inject, args: [CookieStorage,] }] }
];
if (false) {
    /** @type {?} */
    AuthService.prototype.token;
    /** @type {?} */
    AuthService.prototype.storage;
}
