// @gnommostudios/ng-gnommo-base/esm2015/lib/gnommo-base/services/core/base.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { map } from 'rxjs/operators';
import { Injectable, Inject } from '@angular/core';
import { HttpHeaders, HttpClient } from '@angular/common/http';
// unsupported: template constraints.
/**
 * @abstract
 * @template T
 */
export class BaseService {
    /**
     * @param {?} http
     * @param {?} url
     * @param {?} endpoint
     */
    constructor(http, url, endpoint) {
        this.http = http;
        this.url = url;
        this.endpoint = endpoint;
        this.httpOptions = {
            headers: new HttpHeaders({
                // tslint:disable-next-line:max-line-length
                'Accept': '*',
                'Content-type': 'application/json',
            })
        };
    }
    /**
     * @param {?=} headers
     * @return {?}
     */
    getAll(headers) {
        return this.http.get(`${this.url}/${this.endpoint}`, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    getById(_id, headers) {
        return this.http.get(`${this.url}/${this.endpoint}/${_id}`, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    create(item, headers) {
        return this.http.post(`${this.url}/${this.endpoint}`, item, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    update(_id, item, headers) {
        item._id = _id;
        return this.http.patch(`${this.url}/${this.endpoint}`, item, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    updateComplete(_id, item, headers) {
        item._id = _id;
        return this.http.put(`${this.url}/${this.endpoint}`, item, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    delete(_id, headers) {
        return this.http.delete(`${this.url}/${this.endpoint}/${_id}`, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?=} headers
     * @return {?}
     */
    createHttpHeaders(headers) {
        /** @type {?} */
        const httpOptions = Object.assign({}, this.httpOptions);
        if (headers) {
            Object.keys(headers).forEach((headerKey) => {
                httpOptions.headers = httpOptions.headers.append(headerKey, headers[headerKey]);
            });
        }
        return httpOptions;
    }
    /**
     * @param {?=} headers
     * @return {?}
     */
    httpHeadersWithoutAuth(headers) {
        /** @type {?} */
        const httpOptions = {
            headers: new HttpHeaders({
                'Accept': '*',
                'Content-type': 'application/json',
                'X-Skip-Interceptor': ''
            })
        };
        if (headers) {
            Object.keys(headers).forEach((headerKey) => {
                httpOptions.headers = httpOptions.headers.append(headerKey, headers[headerKey]);
            });
        }
        return httpOptions;
    }
}
BaseService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
BaseService.ctorParameters = () => [
    { type: HttpClient, decorators: [{ type: Inject, args: [HttpClient,] }] },
    { type: String, decorators: [{ type: Inject, args: ['url',] }] },
    { type: String, decorators: [{ type: Inject, args: ['endpoint',] }] }
];
if (false) {
    /** @type {?} */
    BaseService.prototype.httpOptions;
    /** @type {?} */
    BaseService.prototype.http;
    /** @type {?} */
    BaseService.prototype.url;
    /** @type {?} */
    BaseService.prototype.endpoint;
}
