// @gnommostudios/ng-gnommo-base/esm2015/lib/gnommo-base/services/core/login-base.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { BaseService } from './base.service';
// unsupported: template constraints.
/**
 * @template T
 */
export class LoginBaseService extends BaseService {
    /**
     * @param {?} http
     * @param {?} url
     * @param {?} endpoint
     */
    constructor(http, url, endpoint) {
        super(http, url, endpoint);
        this.http = http;
        this.url = url;
        this.endpoint = endpoint;
    }
    /**
     * @param {?} credentials
     * @param {?=} headers
     * @return {?}
     */
    login(credentials, headers) {
        return this.http.post(`${this.url}/${this.endpoint}/login`, credentials, this.httpHeadersWithoutAuth(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} email
     * @param {?=} headers
     * @return {?}
     */
    passwordRecovery(email, headers) {
        return this.http.post(`${this.url}/${this.endpoint}/password-recovery`, email, this.httpHeadersWithoutAuth(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} newPassword
     * @param {?} hash
     * @param {?=} headers
     * @return {?}
     */
    resetPassword(newPassword, hash, headers) {
        return this.http.post(`${this.url}/${this.endpoint}/reset-password`, { 'newPassword': newPassword, 'hash': hash }, this.httpHeadersWithoutAuth(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} token
     * @param {?=} headers
     * @return {?}
     */
    getCurrentUser(token, headers) {
        return this.http.get(this.url + '/users-me', this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    register(item, headers) {
        return this.http.post(`${this.url}/${this.endpoint}/register`, item, this.httpHeadersWithoutAuth(headers))
            .pipe(map((data) => data));
    }
}
LoginBaseService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
LoginBaseService.ctorParameters = () => [
    { type: HttpClient, decorators: [{ type: Inject, args: [HttpClient,] }] },
    { type: String, decorators: [{ type: Inject, args: ['url',] }] },
    { type: String, decorators: [{ type: Inject, args: ['endpoint',] }] }
];
if (false) {
    /** @type {?} */
    LoginBaseService.prototype.http;
    /** @type {?} */
    LoginBaseService.prototype.url;
    /** @type {?} */
    LoginBaseService.prototype.endpoint;
}
