// @gnommostudios/ng-gnommo-base/fesm5/gnommostudios-ng-gnommo-base.js
import { Injectable, Inject, NgModule } from '@angular/core';
import { __extends } from 'tslib';
import { map, catchError } from 'rxjs/operators';
import { HttpHeaders, HttpClient, HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { throwError } from 'rxjs';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var CookieHandler = /** @class */ (function () {
    function CookieHandler() {
        this.cookies = {};
    }
    /**
     * @param {?} key
     * @return {?}
     */
    CookieHandler.prototype.get = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (!this.cookies[key]) {
            /** @type {?} */
            var cookie = window.document
                .cookie.split('; ')
                .filter(function (item) { return item.split('=')[0] === key; }).pop();
            if (!cookie) {
                return null;
            }
            this.cookies[key] = this.parse(cookie.split('=').slice(1).join('='));
        }
        return this.cookies[key];
    };
    /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    CookieHandler.prototype.set = /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    function (key, value, expires) {
        this.cookies[key] = value;
        /** @type {?} */
        var cookie = key + "=" + encodeURI(value) + "; path=/" + (expires ? "; expires=" + expires.toUTCString() : '');
        window.document.cookie = cookie;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CookieHandler.prototype.remove = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        document.cookie = key + '=; path=/; expires=Thu, 01 Jan 1970 00:00:01 GMT;';
        delete this.cookies[key];
    };
    /**
     * @param {?} value
     * @return {?}
     */
    CookieHandler.prototype.parse = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        try {
            return JSON.parse(decodeURI(value));
        }
        catch (e) {
            return value;
        }
    };
    CookieHandler.decorators = [
        { type: Injectable },
    ];
    return CookieHandler;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var LocalStorageHandler = /** @class */ (function () {
    function LocalStorageHandler() {
    }
    /**
     * @param {?} key
     * @return {?}
     */
    LocalStorageHandler.prototype.get = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var data = localStorage.getItem(key);
        return this.parse(data);
    };
    /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    LocalStorageHandler.prototype.set = /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    function (key, value, expires) {
        localStorage.setItem(key, typeof value === 'object' ? JSON.stringify(value) : value);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    LocalStorageHandler.prototype.remove = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (localStorage[key]) {
            localStorage.removeItem(key);
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LocalStorageHandler.prototype.parse = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        try {
            return JSON.parse(value);
        }
        catch (e) {
            return value;
        }
    };
    return LocalStorageHandler;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var BaseStorage = /** @class */ (function () {
    function BaseStorage() {
    }
    /**
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.get = /**
     * @param {?} key
     * @return {?}
     */
    function (key) { };
    /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    BaseStorage.prototype.set = /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    function (key, value, expires) { };
    /**
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.remove = /**
     * @param {?} key
     * @return {?}
     */
    function (key) { };
    return BaseStorage;
}());
var CookieStorage = /** @class */ (function (_super) {
    __extends(CookieStorage, _super);
    function CookieStorage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CookieStorage;
}(BaseStorage));
var WebLocalStorage = /** @class */ (function (_super) {
    __extends(WebLocalStorage, _super);
    function WebLocalStorage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return WebLocalStorage;
}(BaseStorage));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// unsupported: template constraints.
/**
 * @abstract
 * @template T
 */
var BaseService = /** @class */ (function () {
    function BaseService(http, url, endpoint) {
        this.http = http;
        this.url = url;
        this.endpoint = endpoint;
        this.httpOptions = {
            headers: new HttpHeaders({
                // tslint:disable-next-line:max-line-length
                'Accept': '*',
                'Content-type': 'application/json',
            })
        };
    }
    /**
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.getAll = /**
     * @param {?=} headers
     * @return {?}
     */
    function (headers) {
        return this.http.get(this.url + "/" + this.endpoint, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.getById = /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    function (_id, headers) {
        return this.http.get(this.url + "/" + this.endpoint + "/" + _id, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.create = /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    function (item, headers) {
        return this.http.post(this.url + "/" + this.endpoint, item, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.update = /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    function (_id, item, headers) {
        item._id = _id;
        return this.http.patch(this.url + "/" + this.endpoint, item, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.updateComplete = /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    function (_id, item, headers) {
        item._id = _id;
        return this.http.put(this.url + "/" + this.endpoint, item, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.delete = /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    function (_id, headers) {
        return this.http.delete(this.url + "/" + this.endpoint + "/" + _id, this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.createHttpHeaders = /**
     * @param {?=} headers
     * @return {?}
     */
    function (headers) {
        /** @type {?} */
        var httpOptions = Object.assign({}, this.httpOptions);
        if (headers) {
            Object.keys(headers).forEach(function (headerKey) {
                httpOptions.headers = httpOptions.headers.append(headerKey, headers[headerKey]);
            });
        }
        return httpOptions;
    };
    /**
     * @param {?=} headers
     * @return {?}
     */
    BaseService.prototype.httpHeadersWithoutAuth = /**
     * @param {?=} headers
     * @return {?}
     */
    function (headers) {
        /** @type {?} */
        var httpOptions = {
            headers: new HttpHeaders({
                'Accept': '*',
                'Content-type': 'application/json',
                'X-Skip-Interceptor': ''
            })
        };
        if (headers) {
            Object.keys(headers).forEach(function (headerKey) {
                httpOptions.headers = httpOptions.headers.append(headerKey, headers[headerKey]);
            });
        }
        return httpOptions;
    };
    BaseService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    BaseService.ctorParameters = function () { return [
        { type: HttpClient, decorators: [{ type: Inject, args: [HttpClient,] }] },
        { type: String, decorators: [{ type: Inject, args: ['url',] }] },
        { type: String, decorators: [{ type: Inject, args: ['endpoint',] }] }
    ]; };
    return BaseService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// unsupported: template constraints.
/**
 * @template T
 */
var LoginBaseService = /** @class */ (function (_super) {
    __extends(LoginBaseService, _super);
    function LoginBaseService(http, url, endpoint) {
        var _this = _super.call(this, http, url, endpoint) || this;
        _this.http = http;
        _this.url = url;
        _this.endpoint = endpoint;
        return _this;
    }
    /**
     * @param {?} credentials
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.login = /**
     * @param {?} credentials
     * @param {?=} headers
     * @return {?}
     */
    function (credentials, headers) {
        return this.http.post(this.url + "/" + this.endpoint + "/login", credentials, this.httpHeadersWithoutAuth(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} email
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.passwordRecovery = /**
     * @param {?} email
     * @param {?=} headers
     * @return {?}
     */
    function (email, headers) {
        return this.http.post(this.url + "/" + this.endpoint + "/password-recovery", email, this.httpHeadersWithoutAuth(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} newPassword
     * @param {?} hash
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.resetPassword = /**
     * @param {?} newPassword
     * @param {?} hash
     * @param {?=} headers
     * @return {?}
     */
    function (newPassword, hash, headers) {
        return this.http.post(this.url + "/" + this.endpoint + "/reset-password", { 'newPassword': newPassword, 'hash': hash }, this.httpHeadersWithoutAuth(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} token
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.getCurrentUser = /**
     * @param {?} token
     * @param {?=} headers
     * @return {?}
     */
    function (token, headers) {
        return this.http.get(this.url + '/users-me', this.createHttpHeaders(headers))
            .pipe(map(function (data) { return data; }));
    };
    /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    LoginBaseService.prototype.register = /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    function (item, headers) {
        return this.http.post(this.url + "/" + this.endpoint + "/register", item, this.httpHeadersWithoutAuth(headers))
            .pipe(map(function (data) { return data; }));
    };
    LoginBaseService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    LoginBaseService.ctorParameters = function () { return [
        { type: HttpClient, decorators: [{ type: Inject, args: [HttpClient,] }] },
        { type: String, decorators: [{ type: Inject, args: ['url',] }] },
        { type: String, decorators: [{ type: Inject, args: ['endpoint',] }] }
    ]; };
    return LoginBaseService;
}(BaseService));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var AuthToken = /** @class */ (function () {
    function AuthToken(data) {
        this.id = null;
        this.ttl = null;
        this.created = null;
        this.userId = null;
        Object.assign(this, data);
    }
    return AuthToken;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var TOKEN_ID = 'token_id';
/** @type {?} */
var TOKEN_TTL = 'token_ttl';
/** @type {?} */
var TOKEN_CREATED = 'token_created';
var AuthService = /** @class */ (function () {
    function AuthService(storage) {
        this.storage = storage;
        this.token = new AuthToken();
        this.token.id = this.storage.get(TOKEN_ID);
        this.token.ttl = this.storage.get(TOKEN_TTL);
        this.token.created = this.storage.get(TOKEN_CREATED);
    }
    /**
     * @return {?}
     */
    AuthService.prototype.getToken = /**
     * @return {?}
     */
    function () {
        return /** @type {?} */ (this.token);
    };
    /**
     * @param {?} tokenId
     * @return {?}
     */
    AuthService.prototype.setToken = /**
     * @param {?} tokenId
     * @return {?}
     */
    function (tokenId) {
        /** @type {?} */
        var token = new AuthToken();
        token.id = tokenId;
        this.persist(TOKEN_ID, token.id, this.expiresTime());
        this.persist(TOKEN_TTL, this.expiresTime(), this.expiresTime());
        this.persist(TOKEN_CREATED, new Date(), this.expiresTime());
        token.ttl = this.expiresTime();
        token.created = new Date();
        this.token = Object.assign({}, this.token, token);
    };
    /**
     * @return {?}
     */
    AuthService.prototype.removeToken = /**
     * @return {?}
     */
    function () {
        this.storage.remove(TOKEN_ID);
        this.storage.remove(TOKEN_TTL);
        this.storage.remove(TOKEN_CREATED);
        this.token = new AuthToken();
    };
    /**
     * @param {?} token_property
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    AuthService.prototype.persist = /**
     * @param {?} token_property
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    function (token_property, value, expires) {
        try {
            this.storage.set("" + token_property, (typeof value === 'object') ? JSON.stringify(value) : value, expires ? expires : null);
        }
        catch (err) {
            // console.error('Cannot access local/session storage:', err);
        }
    };
    /**
     * @return {?}
     */
    AuthService.prototype.expiresTime = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var today = new Date();
        return this.addDays(today, 3);
    };
    /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    AuthService.prototype.addDays = /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    function (date, days) {
        /** @type {?} */
        var result = new Date(date);
        result.setDate(result.getDate() + days);
        return result;
    };
    AuthService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    AuthService.ctorParameters = function () { return [
        { type: CookieStorage, decorators: [{ type: Inject, args: [CookieStorage,] }] }
    ]; };
    return AuthService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var InterceptorSkipHeader = 'X-Skip-Interceptor';
var AuthInterceptor = /** @class */ (function () {
    function AuthInterceptor(authService) {
        this.authService = authService;
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    AuthInterceptor.prototype.intercept = /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    function (req, next) {
        /** @type {?} */
        var token = this.authService.getToken();
        /** @type {?} */
        var authToken = "Bearer " + token.id; // get TOKEN
        // Clone request and update with Auth
        if (req.headers.has(InterceptorSkipHeader)) {
            /** @type {?} */
            var noAuthReq = req.clone({
                headers: req.headers.delete(InterceptorSkipHeader)
            });
            return next.handle(noAuthReq);
        }
        /** @type {?} */
        var authReq = req.clone({
            headers: req.headers.set('Authorization', authToken)
        });
        // send cloned request with header to the next handler.
        return next.handle(authReq);
    };
    AuthInterceptor.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    AuthInterceptor.ctorParameters = function () { return [
        { type: AuthService }
    ]; };
    return AuthInterceptor;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var ErrorInterceptor = /** @class */ (function () {
    function ErrorInterceptor(authService) {
        this.authService = authService;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    ErrorInterceptor.prototype.intercept = /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    function (request, next) {
        var _this = this;
        return next.handle(request).pipe(catchError(function (error) {
            /** @type {?} */
            var errorMsg;
            if (error.status === 401) {
                // auto logout if 401 response returned from api
                // location.reload(true);
                // auto logout if 401 response returned from api
                // location.reload(true);
                _this.authService.removeToken();
                errorMsg = "Error: No tiene autorizaci\u00F3n";
            }
            else {
                errorMsg = error.error.message || error.statusText;
            }
            return throwError(error);
        }));
    };
    ErrorInterceptor.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    ErrorInterceptor.ctorParameters = function () { return [
        { type: AuthService }
    ]; };
    return ErrorInterceptor;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var BaseModel = /** @class */ (function () {
    function BaseModel() {
    }
    return BaseModel;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var GnommoBaseModule = /** @class */ (function () {
    function GnommoBaseModule() {
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    GnommoBaseModule.forRoot = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        return {
            ngModule: GnommoBaseModule,
            providers: [
                BaseService,
                AuthService,
                LoginBaseService,
                // // Interceptors
                { provide: HTTP_INTERCEPTORS, useClass: AuthInterceptor, multi: true },
                { provide: HTTP_INTERCEPTORS, useClass: ErrorInterceptor, multi: true },
                // // Storage
                { provide: WebLocalStorage, useClass: LocalStorageHandler },
                { provide: CookieStorage, useClass: CookieHandler }
            ]
        };
    };
    GnommoBaseModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, HttpClientModule],
                },] },
    ];
    return GnommoBaseModule;
}());
// export * from './interfaces';
// export * from './helpers';
// export * from './models';
// export * from './services';
// export * from './storage';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { CookieHandler, LocalStorageHandler, BaseStorage, WebLocalStorage, CookieStorage, BaseService, LoginBaseService, AuthService, AuthInterceptor, ErrorInterceptor, AuthToken, BaseModel, GnommoBaseModule, AuthInterceptor as ɵg, ErrorInterceptor as ɵh, AuthService as ɵb, BaseService as ɵa, LoginBaseService as ɵf, CookieHandler as ɵj, LocalStorageHandler as ɵi, BaseStorage as ɵc, CookieStorage as ɵd, WebLocalStorage as ɵe };
