// @gnommostudios/ng-gnommo-base/fesm2015/gnommostudios-ng-gnommo-base.js
import { Injectable, Inject, NgModule } from '@angular/core';
import { map, catchError } from 'rxjs/operators';
import { HttpHeaders, HttpClient, HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { throwError } from 'rxjs';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class CookieHandler {
    constructor() {
        this.cookies = {};
    }
    /**
     * @param {?} key
     * @return {?}
     */
    get(key) {
        if (!this.cookies[key]) {
            /** @type {?} */
            const cookie = window.document
                .cookie.split('; ')
                .filter((item) => item.split('=')[0] === key).pop();
            if (!cookie) {
                return null;
            }
            this.cookies[key] = this.parse(cookie.split('=').slice(1).join('='));
        }
        return this.cookies[key];
    }
    /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    set(key, value, expires) {
        this.cookies[key] = value;
        /** @type {?} */
        const cookie = `${key}=${encodeURI(value)}; path=/${expires ? `; expires=${expires.toUTCString()}` : ''}`;
        window.document.cookie = cookie;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    remove(key) {
        document.cookie = key + '=; path=/; expires=Thu, 01 Jan 1970 00:00:01 GMT;';
        delete this.cookies[key];
    }
    /**
     * @param {?} value
     * @return {?}
     */
    parse(value) {
        try {
            return JSON.parse(decodeURI(value));
        }
        catch (e) {
            return value;
        }
    }
}
CookieHandler.decorators = [
    { type: Injectable },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class LocalStorageHandler {
    /**
     * @param {?} key
     * @return {?}
     */
    get(key) {
        /** @type {?} */
        const data = localStorage.getItem(key);
        return this.parse(data);
    }
    /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    set(key, value, expires) {
        localStorage.setItem(key, typeof value === 'object' ? JSON.stringify(value) : value);
    }
    /**
     * @param {?} key
     * @return {?}
     */
    remove(key) {
        if (localStorage[key]) {
            localStorage.removeItem(key);
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    parse(value) {
        try {
            return JSON.parse(value);
        }
        catch (e) {
            return value;
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class BaseStorage {
    /**
     * @param {?} key
     * @return {?}
     */
    get(key) { }
    /**
     * @param {?} key
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    set(key, value, expires) { }
    /**
     * @param {?} key
     * @return {?}
     */
    remove(key) { }
}
class CookieStorage extends BaseStorage {
}
class WebLocalStorage extends BaseStorage {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// unsupported: template constraints.
/**
 * @abstract
 * @template T
 */
class BaseService {
    /**
     * @param {?} http
     * @param {?} url
     * @param {?} endpoint
     */
    constructor(http, url, endpoint) {
        this.http = http;
        this.url = url;
        this.endpoint = endpoint;
        this.httpOptions = {
            headers: new HttpHeaders({
                // tslint:disable-next-line:max-line-length
                'Accept': '*',
                'Content-type': 'application/json',
            })
        };
    }
    /**
     * @param {?=} headers
     * @return {?}
     */
    getAll(headers) {
        return this.http.get(`${this.url}/${this.endpoint}`, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    getById(_id, headers) {
        return this.http.get(`${this.url}/${this.endpoint}/${_id}`, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    create(item, headers) {
        return this.http.post(`${this.url}/${this.endpoint}`, item, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    update(_id, item, headers) {
        item._id = _id;
        return this.http.patch(`${this.url}/${this.endpoint}`, item, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} _id
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    updateComplete(_id, item, headers) {
        item._id = _id;
        return this.http.put(`${this.url}/${this.endpoint}`, item, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} _id
     * @param {?=} headers
     * @return {?}
     */
    delete(_id, headers) {
        return this.http.delete(`${this.url}/${this.endpoint}/${_id}`, this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?=} headers
     * @return {?}
     */
    createHttpHeaders(headers) {
        /** @type {?} */
        const httpOptions = Object.assign({}, this.httpOptions);
        if (headers) {
            Object.keys(headers).forEach((headerKey) => {
                httpOptions.headers = httpOptions.headers.append(headerKey, headers[headerKey]);
            });
        }
        return httpOptions;
    }
    /**
     * @param {?=} headers
     * @return {?}
     */
    httpHeadersWithoutAuth(headers) {
        /** @type {?} */
        const httpOptions = {
            headers: new HttpHeaders({
                'Accept': '*',
                'Content-type': 'application/json',
                'X-Skip-Interceptor': ''
            })
        };
        if (headers) {
            Object.keys(headers).forEach((headerKey) => {
                httpOptions.headers = httpOptions.headers.append(headerKey, headers[headerKey]);
            });
        }
        return httpOptions;
    }
}
BaseService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
BaseService.ctorParameters = () => [
    { type: HttpClient, decorators: [{ type: Inject, args: [HttpClient,] }] },
    { type: String, decorators: [{ type: Inject, args: ['url',] }] },
    { type: String, decorators: [{ type: Inject, args: ['endpoint',] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// unsupported: template constraints.
/**
 * @template T
 */
class LoginBaseService extends BaseService {
    /**
     * @param {?} http
     * @param {?} url
     * @param {?} endpoint
     */
    constructor(http, url, endpoint) {
        super(http, url, endpoint);
        this.http = http;
        this.url = url;
        this.endpoint = endpoint;
    }
    /**
     * @param {?} credentials
     * @param {?=} headers
     * @return {?}
     */
    login(credentials, headers) {
        return this.http.post(`${this.url}/${this.endpoint}/login`, credentials, this.httpHeadersWithoutAuth(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} email
     * @param {?=} headers
     * @return {?}
     */
    passwordRecovery(email, headers) {
        return this.http.post(`${this.url}/${this.endpoint}/password-recovery`, email, this.httpHeadersWithoutAuth(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} newPassword
     * @param {?} hash
     * @param {?=} headers
     * @return {?}
     */
    resetPassword(newPassword, hash, headers) {
        return this.http.post(`${this.url}/${this.endpoint}/reset-password`, { 'newPassword': newPassword, 'hash': hash }, this.httpHeadersWithoutAuth(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} token
     * @param {?=} headers
     * @return {?}
     */
    getCurrentUser(token, headers) {
        return this.http.get(this.url + '/users-me', this.createHttpHeaders(headers))
            .pipe(map((data) => data));
    }
    /**
     * @param {?} item
     * @param {?=} headers
     * @return {?}
     */
    register(item, headers) {
        return this.http.post(`${this.url}/${this.endpoint}/register`, item, this.httpHeadersWithoutAuth(headers))
            .pipe(map((data) => data));
    }
}
LoginBaseService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
LoginBaseService.ctorParameters = () => [
    { type: HttpClient, decorators: [{ type: Inject, args: [HttpClient,] }] },
    { type: String, decorators: [{ type: Inject, args: ['url',] }] },
    { type: String, decorators: [{ type: Inject, args: ['endpoint',] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class AuthToken {
    /**
     * @param {?=} data
     */
    constructor(data) {
        this.id = null;
        this.ttl = null;
        this.created = null;
        this.userId = null;
        Object.assign(this, data);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const TOKEN_ID = 'token_id';
/** @type {?} */
const TOKEN_TTL = 'token_ttl';
/** @type {?} */
const TOKEN_CREATED = 'token_created';
class AuthService {
    /**
     * @param {?} storage
     */
    constructor(storage) {
        this.storage = storage;
        this.token = new AuthToken();
        this.token.id = this.storage.get(TOKEN_ID);
        this.token.ttl = this.storage.get(TOKEN_TTL);
        this.token.created = this.storage.get(TOKEN_CREATED);
    }
    /**
     * @return {?}
     */
    getToken() {
        return /** @type {?} */ (this.token);
    }
    /**
     * @param {?} tokenId
     * @return {?}
     */
    setToken(tokenId) {
        /** @type {?} */
        const token = new AuthToken();
        token.id = tokenId;
        this.persist(TOKEN_ID, token.id, this.expiresTime());
        this.persist(TOKEN_TTL, this.expiresTime(), this.expiresTime());
        this.persist(TOKEN_CREATED, new Date(), this.expiresTime());
        token.ttl = this.expiresTime();
        token.created = new Date();
        this.token = Object.assign({}, this.token, token);
    }
    /**
     * @return {?}
     */
    removeToken() {
        this.storage.remove(TOKEN_ID);
        this.storage.remove(TOKEN_TTL);
        this.storage.remove(TOKEN_CREATED);
        this.token = new AuthToken();
    }
    /**
     * @param {?} token_property
     * @param {?} value
     * @param {?=} expires
     * @return {?}
     */
    persist(token_property, value, expires) {
        try {
            this.storage.set(`${token_property}`, (typeof value === 'object') ? JSON.stringify(value) : value, expires ? expires : null);
        }
        catch (err) {
            // console.error('Cannot access local/session storage:', err);
        }
    }
    /**
     * @return {?}
     */
    expiresTime() {
        /** @type {?} */
        const today = new Date();
        return this.addDays(today, 3);
    }
    /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    addDays(date, days) {
        /** @type {?} */
        const result = new Date(date);
        result.setDate(result.getDate() + days);
        return result;
    }
}
AuthService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
AuthService.ctorParameters = () => [
    { type: CookieStorage, decorators: [{ type: Inject, args: [CookieStorage,] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
const InterceptorSkipHeader = 'X-Skip-Interceptor';
class AuthInterceptor {
    /**
     * @param {?} authService
     */
    constructor(authService) {
        this.authService = authService;
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        /** @type {?} */
        const token = this.authService.getToken();
        /** @type {?} */
        const authToken = `Bearer ${token.id}`; // get TOKEN
        // Clone request and update with Auth
        if (req.headers.has(InterceptorSkipHeader)) {
            /** @type {?} */
            const noAuthReq = req.clone({
                headers: req.headers.delete(InterceptorSkipHeader)
            });
            return next.handle(noAuthReq);
        }
        /** @type {?} */
        const authReq = req.clone({
            headers: req.headers.set('Authorization', authToken)
        });
        // send cloned request with header to the next handler.
        return next.handle(authReq);
    }
}
AuthInterceptor.decorators = [
    { type: Injectable },
];
/** @nocollapse */
AuthInterceptor.ctorParameters = () => [
    { type: AuthService }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class ErrorInterceptor {
    /**
     * @param {?} authService
     */
    constructor(authService) {
        this.authService = authService;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    intercept(request, next) {
        return next.handle(request).pipe(catchError(error => {
            /** @type {?} */
            let errorMsg;
            if (error.status === 401) {
                // auto logout if 401 response returned from api
                // location.reload(true);
                this.authService.removeToken();
                errorMsg = `Error: No tiene autorización`;
            }
            else {
                errorMsg = error.error.message || error.statusText;
            }
            return throwError(error);
        }));
    }
}
ErrorInterceptor.decorators = [
    { type: Injectable },
];
/** @nocollapse */
ErrorInterceptor.ctorParameters = () => [
    { type: AuthService }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class BaseModel {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
class GnommoBaseModule {
    /**
     * @param {?} __0
     * @return {?}
     */
    static forRoot({}) {
        return {
            ngModule: GnommoBaseModule,
            providers: [
                BaseService,
                AuthService,
                LoginBaseService,
                // // Interceptors
                { provide: HTTP_INTERCEPTORS, useClass: AuthInterceptor, multi: true },
                { provide: HTTP_INTERCEPTORS, useClass: ErrorInterceptor, multi: true },
                // // Storage
                { provide: WebLocalStorage, useClass: LocalStorageHandler },
                { provide: CookieStorage, useClass: CookieHandler }
            ]
        };
    }
}
GnommoBaseModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, HttpClientModule],
            },] },
];
// export * from './interfaces';
// export * from './helpers';
// export * from './models';
// export * from './services';
// export * from './storage';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { CookieHandler, LocalStorageHandler, BaseStorage, WebLocalStorage, CookieStorage, BaseService, LoginBaseService, AuthService, AuthInterceptor, ErrorInterceptor, AuthToken, BaseModel, GnommoBaseModule, AuthInterceptor as ɵg, ErrorInterceptor as ɵh, AuthService as ɵb, BaseService as ɵa, LoginBaseService as ɵf, CookieHandler as ɵj, LocalStorageHandler as ɵi, BaseStorage as ɵc, CookieStorage as ɵd, WebLocalStorage as ɵe };
