// @gnommostudios/ng-gnommo-base/esm2015/gnommostudios-ng-gnommo-base.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { CookieHandler, LocalStorageHandler, BaseStorage, WebLocalStorage, CookieStorage, BaseService, LoginBaseService, AuthService, AuthInterceptor, ErrorInterceptor, AuthToken, BaseModel, GnommoBaseModule } from './public_api';
export { AuthInterceptor as ɵg } from './lib/gnommo-base/helpers/auth.interceptor';
export { ErrorInterceptor as ɵh } from './lib/gnommo-base/helpers/error.interceptor';
export { AuthService as ɵb } from './lib/gnommo-base/services/core/auth.service';
export { BaseService as ɵa } from './lib/gnommo-base/services/core/base.service';
export { LoginBaseService as ɵf } from './lib/gnommo-base/services/core/login-base.service';
export { CookieHandler as ɵj } from './lib/gnommo-base/storage/cookie.handler';
export { LocalStorageHandler as ɵi } from './lib/gnommo-base/storage/local-storage.handler';
export { BaseStorage as ɵc, CookieStorage as ɵd, WebLocalStorage as ɵe } from './lib/gnommo-base/storage/storage.handler';
